CIPs/CIP-124/cip-124.md

```diff
@@ -0,0 +1,567 @@
+
+```ts
+struct Neighbor {
+    location: [multiaddr]
+    last_seen: [us_since_epoch]
+    latency: [ms_ping_rtt]
+}
+
+array Neighbors [Neighbor]
+```
+
+A node should probably limit the nodes it responds with to nodes that have be seen recently. e.g. Last 28 days.
+When a new node comes up its existence will get gossiped around the network.
+
+#### **Interests Gossip**
+The Interests are ranges over the recon key space.
+```
+(start, (ahash, cid?), divider, (ahash, cid?), stop)
+```
+
+The messages are sent as CBOR
+```ts
+struct HaveMessage {
+    k: [String], // keys must be 1 longer then hashes unless both are empty
+    h: [String], // hashes must be 1 shorter then keys
+    cid?: [CID], // optional CID of the sub-tree node if remote is a static blob store
+}
+
+struct Interests [
+  {
+    start: String,
+    stop: String,
+    have: HaveMessage,
+  }
+]
+```
+
+## Creating the recon key from a stream.
+
+The key starts with the stream type identifier [varint](https://github.com/multiformats/unsigned-varint).
+
+| name | code | description |
+| --- | --- | --- |
+| Tile | 0 | A stream type representing a Json document |
+| CAIP-10 Link | 1 | Link blockchain accounts to DIDs |
+| Model | 2 | Defines a group of documents in ComposeDB that share a schema |
+| Model Instance Document | 3 | Represents a Json document in ComposeDB |
+| UNLOADABLE | 4 | A stream that is not meant to be loaded |
+
+The following bytes will depend on the stream type
+
+- Tile
+    - `0/<family 6 bytes>/<schema 6 bytes>/<StreamID full>/<Stream Hight [varint](https://github.com/multiformats/unsigned-varint)>/<Event CID full>`
+- CAIP-10 Link
+    - `1/<StreamID full>/<Stream Hight [varint](https://github.com/multiformats/unsigned-varint)>/<Event CID full>`
+- Model
+    - `2/<Model Controller 6 bytes>/<StreamID full>/<Stream Hight [varint](https://github.com/multiformats/unsigned-varint)>/<Event CID full>`
+- Model Instance Document
+    - `3/<Model Controller 6 bytes>/<Model 6 bytes>/<Stream controller 6 bytes><StreamID full>/<Stream Hight [varint](https://github.com/multiformats/unsigned-varint)>/<Event CID full>`
+- UNLOADABLE
+    - Unloadable documents have no structure just use the StreamID as the key
+    - `4/<StreamID>`
+    - `4/kh4q0ozorrgaq2mezktnrmdwleo1d` only unloadable StreamID in [current use](https://www.notion.so/7a8a163fb97047359c56b26b6146ce85)
+
+The main risk in converting the events to keys is that we do want the keys to be grouped in ways that ceramic nodes are most likely to have large contiguous interests like the family for Tile Documents or the Model’s controller for Model Instance Documents. This is in tension with the desire to have short keys. Just the StreamID and event CID could be 124 characters add in an additional 6 char for `Model Controller`, `Model`, `Stream controller`, and 1 char for `Stream Hight` and `stream type` and we are talking about 142 chars for each stream. 
```

Comment:
Make clear that this is the `Stream controller` from the init event and not from replaying the complete stream.

Reply:
Let's update in a subsequent PR.